### PR TITLE
Move packing to trainer and support run directories

### PR DESCRIPTION
**Pull request by Jackmin801**
> [!NOTE]
> Shift batching/packing from orchestrator to trainer via a new transport layer and run_* directories, updating configs, weight broadcast, and tests accordingly.
> 
> - **Core architecture**:
>   - Introduce `prime_rl.transport` (filesystem sender/receiver) and msgspec types (`TrainingSample`, `TrainingBatch`, `MicroBatch`).
>   - Move batching/packing to trainer: new `trainer/rl/packer.py` and `trainer/batch.py`; orchestrator now sends `TrainingBatch` instead of writing micro-batches; remove `orchestrator/batch.py` and `orchestrator/types.py`.
>   - Add `trainer/runs.py` to manage concurrent `run_*` directories and per-run progress; new `utils/pathing.py` for common paths/waits.
> - **Orchestrator**:
>   - Save orch config to `outputs/run_*/configs/orch.toml` and send training batches via `setup_training_batch_sender`.
>   - Update trajectories to use `TrainingSample` with boolean masks.
> - **Trainer**:
>   - Data pipeline reads `TrainingBatch` via transport, packs into micro-batches, and distributes per DP rank.
>   - Weight broadcast (filesystem/NCCL) updated to handle multi-runs and STABLE notifications; optional cleanup per-run.
>   - Move `Progress` to `trainer/runs`; integrate `seq_len` from `trainer.model`.
> - **Configs & validation**:
>   - Add `rollout_transport` to orchestrator/trainer; default `orchestrator.output_dir` to `outputs/run_default`; RL `output_dir` maps to parent with per-run subdir; add `trainer.model.seq_len`.
>   - Validation updated to compare trainer output dir with orchestrator parent; RL auto-setup adjusted; inference adds `max_loras` and vLLM arg mapping.
> - **CI/Docs/Tests**:
>   - CI: run RL LoRA test separately for stability.
>   - Docs: change RL benchmarking flag to `--model.seq-len`.
>   - Tests updated for new types/paths and add `tests/unit/train/test_runs.py`.
>   - Gitignore: ignore `run_*/` and tidy `.idea`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit f5c4b9d704e1ace4fbe8d9bd12a78da7a9032e73. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/transport/__init__.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+from pathlib import Path
+
```
Comment (samsja) [inline src/prime_rl/transport/__init__.py:2]: can we name the folder data_transport

File: src/prime_rl/orchestrator/config.py [hunk 2]
```diff
@@ -441,6 +442,8 @@ class OrchestratorConfig(BaseSettings):
         FileSystemWeightBroadcastConfig()
     )
 
+    transport: Annotated[TransportConfigType, Field(discriminator="type")] = FileSystemTransportConfig()
```
Comment (samsja) [inline src/prime_rl/orchestrator/config.py]:
```suggestion
    data_transport: Annotated[TransportConfigType, Field(discriminator="type")] = FileSystemTransportConfig()
```

File: src/prime_rl/transport/__init__.py (hunk 1 above)
Comment (Jackmin801) [reply src/prime_rl/transport/__init__.py:2]: I think not making the folder specific to rollout / training batch data but transports in general might be nicer

File: src/prime_rl/orchestrator/config.py (hunk 2 above)
Comment (Jackmin801) [reply src/prime_rl/orchestrator/config.py]: maybe rollout transport? data seems a bit general

File: src/prime_rl/trainer/rl/broadcast/nccl.py [hunk 3]
```diff
@@ -155,6 +158,29 @@ def broadcast_weights(self, model: nn.Module, step: int, adapter_only: bool = Fa
         self.logger.debug("Starting broadcasting weights to inference engine via NCCL")
         start_time = time.perf_counter()
         if self.world.is_master:
-            self.notify_orchestrator(step)
+            self.notify_orchestrator()
         self.nccl_broadcast_sender.broadcast_weights(model, step)
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/rl/broadcast/nccl.py:162]: its just the way it works I think

File: src/prime_rl/transport/types.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import msgspec
+
+
+# Orchestrator -> Packer
+class TrainingExample(msgspec.Struct, array_like=True, gc=False, omit_defaults=True):
```
Comment (samsja) [inline src/prime_rl/transport/types.py]:
```suggestion
class TrainingSample(msgspec.Struct, array_like=True, gc=False, omit_defaults=True):
```

I feel example is misleading here. I rpopose to rename to sample.

Comment (Jackmin801) [reply src/prime_rl/transport/types.py]: yea I think that makes sense

File: src/prime_rl/orchestrator/trajectories.py [hunk 5]
```diff
@@ -19,11 +19,11 @@ def interleave_rollout(state: vf.State) -> list[TrainingExample]:
     # Initialize the rollout with prompt and completion from first trajectory step
     trajectory = state["trajectory"]
     first_step = trajectory[0]
-    interleaved_rollout = TrainingExample(
+    interleaved_rollout = TrainingSample(
         prompt_ids=deepcopy(first_step["tokens"]["prompt_ids"]),
-        prompt_mask=deepcopy(first_step["tokens"]["prompt_mask"]),
+        prompt_mask=[i != 0 for i in first_step["tokens"]["prompt_mask"]],
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/trajectories.py]: lol is this to have bool type? if so, can we just call `bool` on it, seems a bit nicer. or we just use the `int` type on the orch side and only turn into bool on trainer when packing into tensor? (then orch type would match vf which i dont dislike as well)

File: src/prime_rl/trainer/rl/broadcast/nccl.py [hunk 6]
```diff
@@ -155,6 +158,29 @@ def broadcast_weights(self, model: nn.Module, step: int, adapter_only: bool = Fa
...
+                if not self.runs.ready_to_update[idx]:
+                    continue
+
+                try:
+                    save_dir = get_step_path(
+                        get_broadcast_dir(self.runs.get_run_dir(idx)), self.runs.progress[idx].step
+                    )
+                    save_dir.mkdir(parents=True, exist_ok=True)
+
+                    # Notify the orchestrator at the end of step to signal that it is safe to load weights from shared filesystem
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/rl/broadcast/nccl.py]: adjust comment

File: src/prime_rl/trainer/rl/broadcast/base.py [hunk 7]
```diff
@@ -5,23 +5,14 @@
...
         self.logger = get_logger()
         self.output_dir = output_dir
-        self.broadcast_dir = get_broadcast_dir(output_dir)
         self.lora_config = lora_config
 
     @abstractmethod
     def broadcast_weights(self, model: nn.Module, step: int):
         pass
-
-    def notify_orchestrator(self, step: int):
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/rl/broadcast/base.py:22]: why does the this method move into the fs and nccl broadcast classes? is it diff for the diff broadcast types now?

File: src/prime_rl/transport/__init__.py
Comment (mikasenghaas) [inline src/prime_rl/transport/__init__.py:1]: wondering abt the position of this module. rn we have 1 module for each of the components of rl run + then shared utils. so by that logic ig should live in `utils.transport`? but also we have been violating this principle alr with e.g. the weight broadcast stuff which is shared logic between trainer and inference which lives on rl trainer rn.. maybe fine to leave it here and cleanup when we rearrange the modules

File: src/prime_rl/orchestrator/config.py [hunk 8]
```diff
@@ -453,7 +456,7 @@ class OrchestratorConfig(BaseSettings):
         Field(
             description="Directory to write outputs to. Will be populated with checkpoints, weights, rollouts and logs as subdirectories. Should be set to a persistent directory with enough disk space. This value should be distinct across experiments running on a single node. See the README for more details."
         ),
-    ] = Path("outputs")
+    ] = Path("outputs/run_default")
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/config.py:459]: does this mean that trainer writes into `outputs/trainer.log` and orch writes into `outputs/run_default/orchestrator.log`? if this is the case do we also need to update tmux.sh or am i missing smth?

**Review by mikasenghaas — COMMENTED** [review]
super well done, love the code structure. we should defo get a multi-run example into our tests. do you think it would make sense to do 2 reverse text or smth? btw,could we run this in regular ci or you think better in nightly tests?

oh also, doing multi-run requires lora right? do we assert this somewhere/ fail loudly if we detect that a config violates this

File: src/prime_rl/orchestrator/config.py (hunk 8 above)
Comment (Jackmin801) [reply src/prime_rl/orchestrator/config.py:459]: Ah yup. good catch!

File: src/prime_rl/transport/__init__.py
Comment (Jackmin801) [reply src/prime_rl/transport/__init__.py:1]: yea imo utils should be lightweight stuff else every component is util which I guess could be what we decide to do. yea maybe separate PR that maybe also moves broadcast

File: src/prime_rl/orchestrator/trajectories.py (hunk 5 above)
Comment (Jackmin801) [reply src/prime_rl/orchestrator/trajectories.py]: I dont think we should b passing around an int for a value that really only accepts bool. bool(i) I guess is ok

File: src/prime_rl/trainer/rl/broadcast/base.py (hunk 7 above)
Comment (Jackmin801) [reply src/prime_rl/trainer/rl/broadcast/base.py:22]: its not meant to be used outside of the class and fs and nccl seem to mean different things with it so I felt like it shouldnt belong in a base class that is defining the general contract of how the Broadcasters should be used

File: src/prime_rl/orchestrator/config.py (hunk 8 above)
Comment (Jackmin801) [reply src/prime_rl/orchestrator/config.py:459]: Ah actually tmux.sh is tailing orchestrator.stdout which is still in outputs/logs so its fine

File: src/prime_rl/trainer/rl/packer.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+    def pack(self):
+        training_batches: dict[int, TrainingBatch] = self.get_batch()
+        start_time = time.time()
+        while not self.has_enough_tokens(training_batches):
+            if time.time() - start_time > TIMEOUT_SECONDS and training_batches:
+                self.logger.warning("Timeout waiting for enough tokens to pack")
+                break
+            time.sleep(1)
+            training_batches = self.get_batch()
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/rl/packer.py:69]: this is intended

File: src/prime_rl/trainer/rl/train.py [hunk 10]
```diff
@@ -153,9 +165,13 @@ def train(config: RLTrainerConfig):
             # Clean up old broadcast directories (unless at ckpt interval if using filesystem weight broadcast)
             ckpt_interval = config.ckpt and config.ckpt.interval
             interval_to_keep = ckpt_interval if config.weight_broadcast.type == "filesystem" else None
-            maybe_clean(weight_broadcast.broadcast_dir, progress.step, config.max_async_level, interval_to_keep)
+            if config.weight_broadcast.type == "filesystem":
+                weight_broadcast.maybe_clean(config.max_async_level, interval_to_keep)
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/rl/train.py:169]: should be fine

File: .github/workflows/gpu_tests.yaml [hunk 11]
```diff
@@ -89,4 +89,6 @@ jobs:
           GITHUB_REF_NAME: ${{ github.ref_name }}
           GITHUB_HEAD_REF: ${{ github.head_ref }}
           PYTEST_OUTPUT_DIR: /tmp/outputs
-        run: uv run pytest tests/integration -m gpu
+        run: |
```
Comment (mikasenghaas) [inline .github/workflows/gpu_tests.yaml:92]: why are we separating those out again? each test should have its own vllm server now no?

File: src/prime_rl/trainer/rl/broadcast/base.py (hunk 7 above)
Comment (mikasenghaas) [reply src/prime_rl/trainer/rl/broadcast/base.py:22]: well i think they mean the same but one notifies that its about to send, while the other notifies that its finished "sending" (writing to disk), but yea prolly fine to keep it in child classes

File: src/prime_rl/trainer/rl/data.py [hunk 12]
```diff
@@ -55,18 +71,33 @@ def _get_micro_batch(self) -> MicroBatch:
...
+                tokenizer=tokenizer,
+                config=config,
+                start_step=start_step,
+            )
+
+        self.receiver: MicroBatchReceiver = setup_micro_batch_receiver(output_dir, self.world.rank, start_step, config)
 
     def wait_for_batch(self) -> None:
-        sync_wait_for_path(self.get_rollout_path())
+        if self.world.is_master:
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/rl/data.py:97]: we pack and then wait?

Comment (mikasenghaas) [reply src/prime_rl/trainer/rl/data.py:97]: would be nice in general to maybe have a descriptive docstring on the dataloader that explains the io that happens here and the interaction with the packer

File: .github/workflows/gpu_tests.yaml (hunk 11 above)
Comment (Jackmin801) [reply .github/workflows/gpu_tests.yaml:92]: the test seems to fail in ci without separating. its not clear to me why. figured better to fix later as its fine to separate them

File: src/prime_rl/trainer/rl/data.py (hunk 12 above)
Comment (Jackmin801) [reply src/prime_rl/trainer/rl/data.py:97]: yep. rank 0 isnt special to the packer so rank 0 also waits for file. we could make packer return 0s batch but i dont imagine it speeds it up that much

Comment (Jackmin801) [reply src/prime_rl/trainer/rl/data.py:97]: yea docstring sounds good. will do

File: .github/workflows/gpu_tests.yaml (hunk 11 above)
Comment (mikasenghaas) [reply .github/workflows/gpu_tests.yaml:92]: ah ok good to know

**Review by samsja — APPROVED** [review]
LFGTM SIR
